apps/studio/src/components/common/texteditor/SQLTextEditor.vue

```diff
@@ -0,0 +1,179 @@
+<template>
+  <text-editor
+    v-bind="$attrs"
+    :value="value"
+    @input="$emit('input', $event)"
+    :lang="lang || 'sql'"
+    :extra-keybindings="keybindings"
+    :hint-options="hintOptions"
+    :columns-getter="columnsGetter"
+    :context-menu-options="handleContextMenuOptions"
+    @initialized="handleInitialized"
+    @paste="handlePaste"
+    @keyup="handleKeyup"
+    @interface="handleInterface"
+  />
+</template>
+
+<script lang="ts">
+import Vue from "vue";
+import TextEditor from "./TextEditor.vue";
+import CodeMirror from "codemirror";
+import { mapState } from "vuex";
+import { registerAutoquote } from "@/lib/codemirror";
+import { removeQueryQuotes } from "@/lib/db/sql_tools";
+import { format } from "sql-formatter";
+import { FormatterDialect, dialectFor } from "@shared/lib/dialects/models";
+
+export default Vue.extend({
+  components: { TextEditor },
+  props: ["value", "lang", "extraKeybindings", "contextMenuOptions"],
+  data() {
+    return {
+      cursorIndex: null,
+      editorInterface: {},
+    };
+  },
+  computed: {
+    ...mapState(["tables"]),
+    hintOptions() {
+      const firstTables = {};
+      const secondTables = {};
+      const thirdTables = {};
+
+      this.tables.forEach((table) => {
+        // don't add table names that can get in conflict with database schema
+        if (/\./.test(table.name)) return;
+
+        // Previously we had to provide a table: column[] mapping.
+        // we don't need to provide the columns anymore because we fetch them dynamically.
+        if (!table.schema) {
+          firstTables[table.name] = [];
+          return;
+        }
+
+        if (table.schema === this.defaultSchema) {
+          firstTables[table.name] = [];
+          secondTables[`${table.schema}.${table.name}`] = [];
+        } else {
+          thirdTables[`${table.schema}.${table.name}`] = [];
+        }
+      });
+
+      const sorted = Object.assign(
+        firstTables,
+        Object.assign(secondTables, thirdTables)
+      );
+
+      return { tables: sorted };
+    },
+    keybindings() {
+      return {
+        "Shift-Ctrl-F": this.formatSql,
+        "Shift-Cmd-F": this.formatSql,
+        ...this.extraKeybindings,
+      };
+    },
+  },
+  methods: {
+    formatSql() {
+      const formatted = format(this.value, {
+        language: FormatterDialect(dialectFor(this.lang)),
+      });
+      this.editorInterface.setValue(formatted);
```

Comment:
same thing here -- we don't wan `editorInterface` leaking up from below. Probably `TextEditor` can handle formatting on it's own, it knows the mime type